### PR TITLE
FEATURE: Cadence 1.0 Core Contracts Changes

### PR DESCRIPTION
**Pull request by joshuahannan**
Includes changes related to Cadence 1.0

Many functions have been changed to `view` for new view function requirements
Updates casting to use correct `as` syntax
changes `pub` to `access(all)`
Adds entitlements
Uses the new fungible token standard and `Burner` contract
Removes custom destructors

### CODE REVIEW

File: contracts/NodeVersionBeacon.cdc
Comment (bluesign) [inline contracts/NodeVersionBeacon.cdc:389]: `var current: UInt64 = 0` can be better maybe

File: contracts/epochs/FlowClusterQC.cdc
Comment (bluesign) [inline contracts/epochs/FlowClusterQC.cdc:108]: is it stable cadence problem ? 

can't be ?  
```
if self.totalWeight == 0 {
                 return 0 
}
 ```

Comment (joshuahannan) [reply contracts/epochs/FlowClusterQC.cdc:108]: Yeah those are just holdovers from older versions of Cadence when we didn't have type inference. I'll change them. Thanks for pointing it out!

**Comment by joshuahannan** [comment]
After many many months, CI is finally passing! Hallelujah!

File: contracts/FlowToken.cdc
Comment (joshuahannan) [reply contracts/FlowToken.cdc:295]: with entitlements, we no longer need to restrict capabilities because now the privledged functions are hidden behind entitlements, so we can link the whole concrete type.

The syntax is invalid because cadence 1.0 remove restricted types, so you can't have a type specification that is `FlowToken.Vault{FungibleToken.Receiver}`. It has to either be `FlowToken.Vault` or `{FungibleToken.Receiver}`

File: contracts/FlowStakingCollection.cdc
Comment (dryruner) [inline contracts/FlowStakingCollection.cdc:1147]: It should be 
`
...
?.check() ?? false
`
here imo - originally it was correct

Comment (joshuahannan) [reply contracts/FlowStakingCollection.cdc:1147]: A recent FLIP is changing `account.capabilities.get` to not return an optional any more
https://github.com/onflow/flips/pull/242
So this is a change to be compatible with that update, which will be released in the next CLI version that should be coming out this week.

Comment (dryruner) [reply contracts/FlowStakingCollection.cdc:1147]: hmm all right

**Comment by joshuahannan** [comment]
I think we're basically ready to merge this to master, so I added a few people who might be able to give it a review! The changes are fairly straightforward, just lots of boilerplate updates to be compatible with the new token standards and Cadence 1.0. Let me know if you'd like to chat with me first about the changes before you review.

**Review by janezpodhostnik — APPROVED** [review]
I mostly checked the changes to the contracts

File: contracts/FlowContractAudits.cdc
Comment (sisyphusSmiling) [inline contracts/FlowContractAudits.cdc:125]: Judging by the `AuditorProxyPublic` interface, looks like there's an `AuditorProxy` Capability published - should these privileged methods then be entitled?

Comment (joshuahannan) [reply contracts/FlowContractAudits.cdc:125]: I probably should just remove this contract. I think we can just let it break because it is not used any more

File: contracts/epochs/FlowClusterQC.cdc
Comment (sisyphusSmiling) [inline contracts/epochs/FlowClusterQC.cdc:275]: Should this be replaced with `Burner.burnCallback()`

**Review by sisyphusSmiling — COMMENTED** [review]
Leaving a few follow up questions about entitlements. They're probably alright, I just lack the context to understand if there's any risk of down casting from public capabilities.

File: contracts/FlowContractAudits.cdc
Comment (joshuahannan) [reply contracts/FlowContractAudits.cdc:1]: I'm removing this contract

File: contracts/FlowServiceAccount.cdc
Comment (joshuahannan) [reply contracts/FlowServiceAccount.cdc:28]: Good suggestion. Always good to follow best practices

File: contracts/FlowStakingCollection.cdc
Comment (joshuahannan) [reply contracts/FlowStakingCollection.cdc:1161]: Yeah, probably should be. I'll need to think about it

File: contracts/LockedTokens.cdc
Comment (joshuahannan) [reply contracts/LockedTokens.cdc:102]: great catch! I was focused on `TokenHolder` so i forgot that this had public capabilities also. 😅 I'm updating this

File: contracts/FlowToken.cdc
Comment (joshuahannan) [reply contracts/FlowToken.cdc:80]: good point. I added comments

File: contracts/LockedTokens.cdc
Comment (joshuahannan) [reply contracts/LockedTokens.cdc:148]: yeah, good catch

**Comment by joshuahannan** [comment]
@sisyphusSmiling The changes were big enough that I just created a PR with them for you to review. Can you take a look?
https://github.com/onflow/flow-core-contracts/pull/429

File: contracts/FlowIDTableStaking.cdc
Comment (sisyphusSmiling) [inline contracts/FlowIDTableStaking.cdc:267]: Doesn't seem to be a problem since the core contracts have been migrated multiple times, but I'm surprised this is a compatible migration. A reference field would mean that the struct isn't storable, wouldn't it?

File: contracts/FlowIDTableStaking.cdc
Comment (sisyphusSmiling) [inline contracts/FlowIDTableStaking.cdc:421]:
```suggestion
        access(self) view fun isEligibleForCandidateNodeStatus(_ nodeRecord: &FlowIDTableStaking.NodeRecord): Bool {
```

File: contracts/FlowIDTableStaking.cdc
Comment (sisyphusSmiling) [inline contracts/FlowIDTableStaking.cdc:1026]: Seems like this iteration and the logic within it is repeated a couple times. Could this be made into a helper or maybe an iterator like `forEachDelegator`? Not pressing at all, maybe something to think about as a change to include later.

File: transactions/FlowServiceAccount/set_execution_effort_weights.cdc
Comment (sisyphusSmiling) [inline transactions/FlowServiceAccount/set_execution_effort_weights.cdc:42]: Applies to all transactions where raw values are saved in storage - do we need any sort of input validation for these? I'd imagine the values are sensitive and probably live in storage due to contract update limitations, but adding validation in the committing transaction would at least *feel* a bit safer.

Getting out of scope a bit, but if we know what these values will remain and don't think more will be added in the future, we could move them to some kind of config resource or contract and add setters so the input validation is done at the contract level and we could emit events when the values are updated.

**Review by sisyphusSmiling — APPROVED** [review]
Awesome work! Just some smaller suggestions and comments.

File: contracts/FlowIDTableStaking.cdc
Comment (joshuahannan) [reply contracts/FlowIDTableStaking.cdc:267]: nobody stores this struct so it isn't a problem. It just gives a snapshot of a node's stake at a specific point in time so storing it wouldn't make sense.

File: contracts/FlowIDTableStaking.cdc
Comment (joshuahannan) [reply contracts/FlowIDTableStaking.cdc:1026]: Good idea! I created an issue for later: https://github.com/onflow/flow-core-contracts/issues/431

**Comment by joshuahannan** [comment]
@jordanschalm Would you be able to give a review of just the contracts here? No need to look at all the tests and transactions since that is mostly just boilerplate. Want to get one more approval on this before we merge it.

File: contracts/LockedTokens.cdc
Comment (jordanschalm) [inline contracts/LockedTokens.cdc:669]:
```suggestion
     access(all) entitlement AccountCreator

     access(all) resource interface AddAccount {
          access(AccountCreator) fun addAccount(
```
spacing is off here

File: contracts/epochs/FlowEpoch.cdc
Comment (jordanschalm) [inline contracts/epochs/FlowEpoch.cdc:48]: Dang, every case in an enum needs its own access modifier? 😮

**Review by jordanschalm — COMMENTED** [review]
Took a pass over the **contract files only**. Mainly noted a few questions and places where access modifiers could potentially be tightened.

File: contracts/FlowIDTableStaking.cdc
Comment (joshuahannan) [reply contracts/FlowIDTableStaking.cdc:365]: Cadence 1.0 is [removing custom destructors](https://github.com/onflow/flips/blob/main/cadence/20230811-destructor-removal.md). Resources are still able to be destroyed, but now a resource can't execute custom logic when it is destroyed unless it implements the `Burner.Burnable` [interface](https://github.com/onflow/flow-ft/blob/master/contracts/utility/Burner.cdc#L16). 

Resources that are owned by other resources just get destroyed if the owning resource is destroyed

File: contracts/FlowIDTableStaking.cdc
Comment (joshuahannan) [reply contracts/FlowIDTableStaking.cdc:407]: It just means that the type definition is accessible to the public. If it wasn't `access(all)`, then you couldn't use it to create an entitled capability in a transaction because it wouldn't be accessible to the transaction

File: contracts/FlowIDTableStaking.cdc
Comment (joshuahannan) [reply contracts/FlowIDTableStaking.cdc:1906]: Good catch. I'll add this

File: contracts/FlowIDTableStaking.cdc
Comment (joshuahannan) [reply contracts/FlowIDTableStaking.cdc:1917]: I'll add this too

File: contracts/FlowStakingCollection.cdc
Comment (joshuahannan) [reply contracts/FlowStakingCollection.cdc:972]: No, they use `append`, which isn't allowed in view functions. We could refactor the code to do it without append, but I need to focus on other work right now.

File: contracts/FlowToken.cdc
Comment (joshuahannan) [reply contracts/FlowToken.cdc:119]: I need to double check with Vishal about that, but we had decided that we didn't want to remove it as part of Crescendo because there were partners who it would be a breaking change for. I still would prefer to remove it since there are other breaking changes. I'll ask again. I created an issue to track it: https://github.com/onflow/flow-core-contracts/issues/434

File: contracts/epochs/FlowEpoch.cdc
Comment (joshuahannan) [reply contracts/epochs/FlowEpoch.cdc:48]: yeah, a little awkward

**Comment by joshuahannan** [comment]
Thank you so much for the review @jordanschalm! I'll work on getting all the Cadence dependencies updated for this and the other contract repos soon and then we'll be able to merge this!

**Comment by joshuahannan** [comment]
Finally merged to master! Thank you everyone for all the reviews and help on this! 🥳 🎉

**Pull request merged**
